core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1830,6 +1831,19 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     }
   }
 
+  def checkSamActionWithLock(workspaceName: WorkspaceName, samAction: SamResourceAction): Future[PerRequestMessage] = {
+    val testFuture = dataSource.inTransaction { dataAccess =>
+      withWorkspaceContext(workspaceName, dataAccess, Some(WorkspaceAttributeSpecs(all=false))) { workspaceContext =>
+        requireAccess(workspaceContext.workspace, samAction) {
+          DBIO.successful(RequestComplete(StatusCodes.NoContent)) //if we get here, we passed all the hoops
+        }
+      }
+    }
+    //if we failed for any reason, the user can't do that thing on the workspace
+    testFuture.recover { case _ =>
+      RequestComplete(StatusCodes.Unauthorized) }
+  }
```

Comment:
I know it's rampant throughout Rawls, but let's not add more: could you rewrite this so we're not making REST calls  within a db transaction?

Reply:
sure